Implement mounting and unmounting for external journal device

jason50123 opened the pull request:
This pull request continues from https://github.com/sysprog21/simplefs/pull/59#issue-2363416768. 
Due to some issues with Git, I have addressed them and created this new pull request.


Added a function simplefs_parse_options in the fill super process to obtain the external journal device and configure it using jbd2 related functions. Additionally, journal-related functionality was added to the write functions to record metadata "extent" information.

Also, added a "make journal" command in the Makefile, allowing users to create an external journal device image and mount it to simplefs.

jason50123 committed:
Implement External Journal Device Initialization

This update introduces the following changes:

1. Added a function simplefs_parse_options in the fill super process
to obtain the external journal device and configure it using jbd2
related functions. This allows the file system to correctly initialize
and manage the external journal device.

2. Implemented additional functions in put_super to handle the journal
device during the unmount process. This ensures that the journal device
is properly managed and released when the file system is unmounted.

3. Added a make journal command in the Makefile. This new command allows
 users to create an external journal device image and mount it to
 simplefs, simplifying the setup and usage of the external journal.

Future Work:
Currently, the external journal device size is fixed at 8MB. To use a
different size, corresponding variables in the code need to be modified.
Additionally, support for an internal journal (inode journal) can be
added in the future to further improve the file system's capabilities.

jserv commented on super.c:
```diff
@@ -196,6 +201,321 @@ static int simplefs_statfs(struct dentry *dentry, struct kstatfs *stat)
+        errno = -EINVAL;
+        goto out_bdev;
+    }
+    // sbi = (struct simplefs_sb_info *) (bh->b_data + offset);
```
Don't leave the dead code.

jserv commented:
CI pipeline complains the following:
```
Testing cmd: touch 40919.txt...
Testing cmd: touch 40920.txt...Unmounting filesystem...
Remounting filesystem...
Remount succeeds.
Failed: file_1.txt content is incorrect.
Error: Process completed with exit code 2.
```

jserv commented:
> This pull request continues from [#59 (comment)](https://github.com/sysprog21/simplefs/pull/59#issue-2363416768). Due to some issues with Git, I have addressed them and created this new pull request.

Don't do this. You should simply run `git rebase -i` and then `git push --force`.
If a workman wishes to do a good job, he must first sharpen his tools.

jason50123 commented:
> CI pipeline complains the following:
> 
> ```
> 
> Testing cmd: touch 40919.txt...
> 
> Testing cmd: touch 40920.txt...Unmounting filesystem...
> 
> Remounting filesystem...
> 
> Remount succeeds.
> 
> Failed: file_1.txt content is incorrect.
> 
> Error: Process completed with exit code 2.
> 
> ```

@jserv This version has not yet addressed issues related to Put super. So should I handle this part first and then submit a pull request?

jason50123 commented:
This version has not yet addressed issues related to Put super. So should I handle this part first and then submit a pull request?
 
> CI pipeline complains the following:
> 
> ```
> 
> Testing cmd: touch 40919.txt...
> 
> Testing cmd: touch 40920.txt...Unmounting filesystem...
> 
> Remounting filesystem...
> 
> Remount succeeds.
> 
> Failed: file_1.txt content is incorrect.
> 
> Error: Process completed with exit code 2.
> 
> ```

jason50123 commented:
> > @jserv This version has not yet addressed issues related to Put super. So should I handle this part first and then submit a pull request?
> 
> 
> 
> Up to you. You should resolve the issue reported by GitHub Actions in advance. Later, you can create a new commit within the pull request (again, using `git rebase -i`) or submit a new pull request to consolidate the CI pipeline.

I understand, I will take care of it.

jserv reviewed (changes requested):
Use `git rebase` rather than `git merge`!

jserv commented on Makefile:
```diff
@@ -10,6 +10,9 @@ all: $(MKFS)
 IMAGE ?= test.img
 IMAGESIZE ?= 200
+JOURNAL ?= journal.img
+JOURNALSIZE ?= 8
```
How did you calculate for journal size?

jason50123 replied:
Assuming an 8 MB size for the external journal device, which is an arbitrary choice for now, I will set the journal block length to a fixed 2048, calculated by dividing the device size by the block size (4096 bytes).

jserv replied:
Mention the above in `README.md` as well.

jserv reviewed (changes requested):
Update top-level `README.md` for explaining journaling support.

jserv commented on README.md:
````diff
@@ -172,6 +172,50 @@ struct simplefs_extent
                                  +---------+
 ```
 
+### journalling support
````
Add some descriptive text here, such as the relationship with jbd2.

jserv commented on README.md:
````diff
@@ -172,6 +172,74 @@ struct simplefs_extent
+```
+
+
 ## TODO
````
Remove the TODO here.

jserv commented on super.c:
```diff
@@ -331,6 +719,6 @@ int simplefs_fill_super(struct super_block *sb, void *data, int silent)
 release:
     brelse(bh);
-
+    pr_err("simplefs_fill_super: failed to load superblock\n");
```
Don't put `pr_err` here.

jserv commented on super.c:
```diff
@@ -319,6 +696,17 @@ int simplefs_fill_super(struct super_block *sb, void *data, int silent)
+        goto release;
+    }
+
+    pr_info("simplefs_fill_super: successfully loaded superblock\n");
```
Remove `pr_info` here.

jserv commented on super.c:
```diff
@@ -196,6 +244,335 @@ static int simplefs_statfs(struct dentry *dentry, struct kstatfs *stat)
+            break;
+
+        case SIMPLEFS_OPT_JOURNAL_PATH: {
+            char *journal_path;
```
Move the declarations out of switch-case.

jserv commented:
Can you provide journal specific test cases?

jason50123 commented:
> Can you provide journal specific test cases?

@jserv  Due to certain issues, unexpected errors occur after journal data during mount-remount operations. Therefore, this commit only addresses the mounting and unmounting of the external journal device. The journaling metadata part will be submitted in a pull request after thorough testing is completed.

jserv commented:
> Due to certain issues, unexpected errors occur after journal data during mount-remount operations. Therefore, this commit only addresses the mounting and unmounting of the external journal device. The journaling metadata part will be submitted in a pull request after thorough testing is completed.

Then, you should consolidate the git commit message and address the known issue. We do still need some preliminary test cases.

jserv reviewed (changes requested):
For the sake of Linux API changes, there are several implementation of `simplefs_get_dev_journal` function. Move some common code in a helper function, which is independent from Linux kernel versions, so that we can shrink the lines of code.

jserv commented on super.c:
```diff
@@ -331,6 +635,5 @@ int simplefs_fill_super(struct super_block *sb, void *data, int silent)
     kfree(sbi);
 release:
     brelse(bh);
-
```
Don't do this.

jason50123 replied:
I've made the following changes: when simplefs_parse_options fails, it will directly return the error code (ret), instead of using goto release

jserv replied:
You should keep the original blank line. Minimize the necessary changes.

jserv commented on super.c:
```diff
@@ -196,6 +244,253 @@ static int simplefs_statfs(struct dentry *dentry, struct kstatfs *stat)
+    }
+
+    /*
+     * FIXME:
```
Improve this comment to address the problem you are facing and place the possible solution for future work.

jserv commented on super.c:
```diff
@@ -109,6 +114,49 @@ static int simplefs_write_inode(struct inode *inode,
+    }
+#endif
+
+
```
Remove this blank line.

jserv commented on README.md:
```diff
@@ -172,9 +172,71 @@ struct simplefs_extent
+Simplefs now includes support for an external journal device, leveraging the journaling block device (jbd2) subsystem in the Linux kernel. This enhancement improves the file system's resilience by maintaining a log of changes, which helps prevent corruption and facilitates recovery in the event of a crash or power failure.
+
+
+The journaling support in simplefs is implemented using the jbd2 subsystem, which is a widely-used journaling layer in Linux. Currently, simplefs primarily stores the journal-related information in an external journal device.
```
Mention the current limitation and known issues for journaling support.

jserv commented on README.md:
```diff
@@ -172,9 +172,85 @@ struct simplefs_extent
+
+1. External Journal Device Size:
+
+  - Currently, the exact size of the external journal device cannot be determined. As a temporary solution, the size is set by dividing the device size by the block size, with the external journal device size fixed at 8 MB.
```
Avoid repeated "Currently." Refine the text to be more concrete.

jserv commented on super.c:
```diff
@@ -109,6 +114,48 @@ static int simplefs_write_inode(struct inode *inode,
+        sync_blockdev(sbi->s_journal_bdev_handle->bdev);
+        invalidate_bdev(sbi->s_journal_bdev_handle->bdev);
+    }
+/* simplefs.h? */
```
What did you mean?

jason50123 replied:
I just forgot to delete it

jserv replied:
It still exists.

jserv commented on super.c:
```diff
@@ -196,6 +243,256 @@ static int simplefs_statfs(struct dentry *dentry, struct kstatfs *stat)
     return 0;
 }
 
+/* journal relation code */
```
Improve the comment.

jserv commented on super.c:
```diff
@@ -196,6 +243,256 @@ static int simplefs_statfs(struct dentry *dentry, struct kstatfs *stat)
+
+    journal->j_private = sb;
+    return journal;
+
```
Remove a blank line.

jserv commented on super.c:
```diff
@@ -196,6 +243,256 @@ static int simplefs_statfs(struct dentry *dentry, struct kstatfs *stat)
+                pr_err("simplefs_parse_options: match_int failed\n");
+                return 1;
+            }
+            printk(KERN_INFO "Loading journal devnum: %i\n", arg);
```
Remove this line since we can catch the fatal case.

jserv commented on super.c:
```diff
@@ -319,6 +616,15 @@ int simplefs_fill_super(struct super_block *sb, void *data, int silent)
+    ret = simplefs_parse_options(sb, data);
+    if (ret) {
+        pr_err(
+            "simplefs_fill_super: simplefs_parse_options failed with error "
```
Improve the error message.

jserv reviewed (changes requested):
Improve the git commit messages to consist of several complete English sentences.

jserv commented:
Thank @jason50123 for contributing!
